exercises/book/A3-traits-generics/mod.md

```diff
@@ -0,0 +1,151 @@
+- `insert` inserts an item at the given index and increments the length of the `LocalStorageVec`
+- `remove` removes an item at the given index and returns it.
+- `clear` resets the length of the `LocalStorageVec` to 0.
+
+ Uncomment the corresponding test cases and make them compile and pass. **Be sure to have a look at the methods provided for slices [`[T]`](https://doc.rust-lang.org/std/primitive.slice.html) and [`Vec<T>`](https://doc.rust-lang.org/std/vec/struct.Vec.html)** Specifically, `[T]::copy_within` and `Vec::extend_from_slice` can be of use.
+
+
+### A3.2.D `Iterator` and `IntoIterator`
+Our `LocalStorageVec` can be used in the real world now, but we still shoudn't be statisfied. There are various traits in the standard library that we can implement for our `LocalStorageVec` that would make users of our crate happy. 
+
+First off, we will implement the [`IntoIterator`](https://doc.rust-lang.org/std/iter/trait.IntoIterator.html) and [`Iterator`](https://doc.rust-lang.org/std/iter/trait.Iterator.html) traits. Go ahead and uncomment the `it_iters` test case. Let's define a new type:
+
+
+```rust
+pub struct LocalStorageVecIter<T, const N: usize> {
+    vec: LocalStorageVec<T, N>,
+    counter: usize,
+}
+```
+
+This is the type we'll implement the `Iterator` trait on. You'll need to specify the item this `Iterator` implementation yields, as well as an implementation for `Iterator::next`, which yields the next item. You'll be able to make this easier by bounding `T` to `Default` when implementing the `Iterator` trait, as then you can use the [`std::mem::take`](https://doc.rust-lang.org/std/mem/fn.take.html) function to take an item from the `LocalStorageVec` and replace it with the default value for `T`.
+
+Take a look at the list of methods under the ['provided methods' section](https://doc.rust-lang.org/std/iter/trait.Iterator.html). In there, lots of useful methods that come free with the implementation of the `Iterator` trait are defined, and implemented in terms of the `next` method. Knowing in the back of your head what methods there are, greatly helps in improving your efficiency in programming with Rust. Which of the provided methods can you override in order to make the implementation of `LocalStorageVecIter` more efficient, given that we can access the fields and methods of `LocalStorageVec`?
+
+Now to instantiate a `LocalStorageVecIter`, implement the [`IntoIter`] trait for it, in such a way that calling `into_iter` yields a `LocalStorageVecIter`.
+
+### A3.2.E `AsRef` and `AsMut`
+[`AsRef`](https://doc.rust-lang.org/std/convert/trait.AsRef.html) and [`AsMut`](https://doc.rust-lang.org/std/convert/trait.AsMut.html) are used to implement *cheap* reference-to-reference coercion. For instance, our `LocalStorageVec<T, N>` is somewhat similar to a slice `&[T]`, as both represent a contiguous series of `T` values. This is true whether the `LocalStorageVec` buffer resides on the stack or on the heap. 
+
```

Comment:
Note the added test-case in PR #35; it's very "easy" to forget to limit the &[T] returned by this method in case the vector is located on the stack. Maybe some words about it can be added here as well.

Reply:
Good catch!